genie-web/src/main/java/com/netflix/genie/web/scripts/ManagedScript.java

```diff
@@ -0,0 +1,95 @@
+/*
+ *
+ *  Copyright 2019 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.scripts;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.genie.web.exceptions.checked.ScriptExecutionException;
+import com.netflix.genie.web.exceptions.checked.ScriptNotConfiguredException;
+import io.micrometer.core.instrument.MeterRegistry;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.PostConstruct;
+import javax.script.Bindings;
+import javax.script.SimpleBindings;
+import java.net.URI;
+import java.util.Map;
+
+/**
+ * Abstract script class for components that rely on an external script to be loaded and invoked at runtime.
+ * This class makes use of {@link ScriptManager} to execute.
+ * <p>
+ * TODO: add metrics
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+abstract class ManagedScript {
+    protected final ObjectMapper mapper;
+    private final ScriptManager scriptManager;
+    private final ManagedScriptBaseProperties properties;
+    private final MeterRegistry registry;
+
+    protected ManagedScript(
+        final ScriptManager scriptManager,
+        final ManagedScriptBaseProperties properties,
+        final ObjectMapper mapper,
+        final MeterRegistry registry
+    ) {
+        this.scriptManager = scriptManager;
+        this.properties = properties;
+        this.mapper = mapper;
+        this.registry = registry;
+    }
+
+    @PostConstruct
+    void loadPostConstruct() {
+        final URI scriptUri = this.properties.getSource();
+        if (this.properties.isAutoLoadEnabled() && scriptUri != null) {
+            this.scriptManager.manageScript(scriptUri);
+        }
+    }
+
+    protected Object evaluateScript(
```

Comment:
this may not be pertinent to this PR but part of me wonders if we should standardize on one type of script (`groovy` for example) rather than just supporting any kind of script. This might allow us to provide a better contract here for scripts. I'm a little concerned that this abstracts it to such a level that there's really no contract that we can enforce. It might be nice in the future to have say "ok if you want a `ClusterSelector` script here are your input object types and the expected output type is a `Cluster` or something concrete instead of just Object.

I mean I guess I could see it being super generic here (kind of pythonic) but we do lose something by doing that while gaining the reusability. Maybe templetize `ManagedScript` somehow?

Reply:
Right now that contract is one layer above in concrete `*Script` class that extends the base `ManagedScript`. Users of the latter pass typed parameters and receive a result of the expected type.

But i agree we if we could reduce support to just Groovy, we could do some nicer things.